Allow interactive commands and return correct exit code

jfahrer opened the pull request:
This will
1) forward the output to stdout before the command finishes
2) allow starting interactive applications like a shell

elsesiy commented on donner.go:
```diff
@@ -105,15 +106,21 @@ func execCommand(cfg *Cfg, params []string) error {
+	cmd.Stderr = os.Stderr
+
+	var waitStatus syscall.WaitStatus
+	if err := cmd.Run(); err != nil {
```
This whole block handles only the error case which we technically don't need because docker is logging to `stderr` which you got covered by assigning `cmd.Stderr = os.Stderr`. The only case you're handling with the waitstatus is when the command can't be executed which could be as simple as L115.

elsesiy commented on donner.go:
```diff
@@ -105,15 +106,21 @@ func execCommand(cfg *Cfg, params []string) error {
+			os.Exit(waitStatus.ExitStatus())
+		}
+	} else {
+		os.Exit(waitStatus.ExitStatus())
```
This doesn't make sense to me. Shouldn't we be looking at `cmd.ProcessState` to determine if the command finished successfully / error'd out? I think we can vastly simplify this snippet unless I'm overseeing something...

jfahrer commented:
I simplified it a little bit and removed one of the branches in the condition.

What I wanted to achieve:
- make sure we get the output of whatever we run right away
- we have all streams attached so that we see errors
- enable running interactive applications like a shell
- ensure we exit with the same error code as the command we executed
- deal with the fact that go might not be able to run the command at all (permissions, missing, ...)

My understanding is that `cmd.Run()` will return an `ExitError` if the command returns an exitcode != 0. In this case, we want to return the same error. Hence the 
```
waitStatus := exitError.Sys().(syscall.WaitStatus)
wos.Exit(waitStatus.ExitStatus())
```
spiel.

`cmd.Run()` will return other errors for cases where go couldn't start the child process at all.

Does that make sense?

jfahrer commented:
```
strategies:
  run:
    handler: docker_run
    image: alpine:3.6
  broken:
    handler: missing_executable

default_strategy: run

commands:
  sh: run
  foo: broken
```

^^ example config to test this:

```
donner run sh -c 'exit 0'
echo $?
donner run sh -c 'exit 2'
echo $?
donner run foo
echo $?
```

jfahrer commented:
Merging this in, we can optimize later.
